code/datums/components/anti_magic.dm

```diff
@@ -1,54 +1,148 @@
+ * * drain_antimagic (optional) The proc that is triggered when an object has been drained a antimagic charge
+ * * expiration (optional) The proc that is triggered when the object is depleted of charges
+ * *
+ * antimagic bitflags: (see code/__DEFINES/magic.dm)
+ * * MAGIC_RESISTANCE - Default magic resistance that blocks normal magic (wizard, spells, staffs)
+ * * MAGIC_RESISTANCE_MIND - Tinfoil hat magic resistance that blocks mental magic (telepathy, abductors, jelly people)
+ * * MAGIC_RESISTANCE_HOLY - Holy magic resistance that blocks unholy magic (revenant, cult, vampire, voice of god)
+**/
+/datum/component/anti_magic/Initialize(
+		antimagic_flags = MAGIC_RESISTANCE,
+		charges = INFINITY,
+		inventory_flags = ~ITEM_SLOT_BACKPACK, // items in a backpack won't activate, anywhere else is fine
+		datum/callback/drain_antimagic,
+		datum/callback/expiration
+	)
+
 	if(isitem(parent))
 		RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/on_equip)
 		RegisterSignal(parent, COMSIG_ITEM_DROPPED, .proc/on_drop)
 	else if(ismob(parent))
-		RegisterSignal(parent, COMSIG_MOB_RECEIVE_MAGIC, .proc/protect)
+		RegisterSignal(parent, COMSIG_MOB_RECEIVE_MAGIC, .proc/block_receiving_magic)
+		RegisterSignal(parent, COMSIG_MOB_RESTRICT_MAGIC, .proc/restrict_casting_magic)
+		to_chat(parent, span_warning("Magic seems to flee from you. You are immune to spells but are unable to cast magic."))
 	else
 		return COMPONENT_INCOMPATIBLE
 
-	magic = _magic
-	holy = _holy
-	psychic = _psychic
-	if(_allowed_slots)
-		allowed_slots = _allowed_slots
-	if(!isnull(_charges))
-		charges = _charges
-	blocks_self = _blocks_self
-	reaction = _reaction
-	expire = _expire
+	src.antimagic_flags = antimagic_flags
+	src.charges = charges
+	src.inventory_flags = inventory_flags
+	src.drain_antimagic = drain_antimagic
+	src.expiration = expiration
+
+/datum/component/anti_magic/Destroy(force, silent)
+	QDEL_NULL(drain_antimagic)
+	QDEL_NULL(expiration)
+	return ..()
 
 /datum/component/anti_magic/proc/on_equip(datum/source, mob/equipper, slot)
 	SIGNAL_HANDLER
 
-	if(!(allowed_slots & slot)) //Check that the slot is valid for antimagic
+	if(!(inventory_flags & slot)) //Check that the slot is valid for antimagic
 		UnregisterSignal(equipper, COMSIG_MOB_RECEIVE_MAGIC)
+		UnregisterSignal(equipper, COMSIG_MOB_RESTRICT_MAGIC)
+		equipper.update_action_buttons()
 		return
-	RegisterSignal(equipper, COMSIG_MOB_RECEIVE_MAGIC, .proc/protect, TRUE)
+	RegisterSignal(equipper, COMSIG_MOB_RECEIVE_MAGIC, .proc/block_receiving_magic)
+	RegisterSignal(equipper, COMSIG_MOB_RESTRICT_MAGIC, .proc/restrict_casting_magic)
+	equipper.update_action_buttons()
+
+	// Check to see if we have any spells that are blocked due to antimagic
+	for(var/obj/effect/proc_holder/spell/magic_spell in equipper.mind?.spell_list)
+		if(antimagic_flags & magic_spell.antimagic_flags)
+			to_chat(equipper, span_warning("[parent] is interfering with your ability to cast magic!"))
```

Comment:
This message gets really annoying, as it's sent every time it's equipped.

Equipped includes picking it up, putting it in your pocket or belt or suit storage...

Reply:
Fixed with https://github.com/tgstation/tgstation/pull/64124/commits/9941f8c4bb77e17b04edf84ef71d88863453f75c